Containerfile

```diff
@@ -1,18 +1,20 @@
-FROM registry.fedoraproject.org/fedora-minimal:39
+# hadolint global ignore=DL3007
+FROM registry.access.redhat.com/ubi9/ubi-minimal:latest
```

Comment:
To have reproducible and tested builds, we should avoid to have `latest` tag, the best option would be to have floating tag with pinned digest and make dependabot to update it regularly. Also latest tag in ubi-minimal will cause to bump minor versions, they sometimes contain not compatible changes.

Reply:
Finally, someone else who sees the evil of `:latest` :grin:

Reply:
The way you're presenting the "evilness" of latest in context of solving it with digests and dependabot is a bit misleading and maybe even false advertisement. Now, version locks monitored by dependabot make sense with individual packages because those are enclosed logical units where you always get **that** particular version. It doesn't work that way with container images, unless you never run an install on the package manager inside. The reason for that is that now we're talking about a set of packages that may get updated without you realizing and still thinking that you're using your safe pinned container image. The problem arises the moment you run e.g. `dnf install` because it'll likely update a subset of the base packages without you explicitly asking for it. Here's an example of it:

This is snippet of RockyLinux cachi2 container install with a newer (sha256:4b9c7b0285aa68f45ef42308fb7a364ffb63b461645cfb22c53212d85d16bf93) digest
```
================================================================================
 Package                 Arch    Version                       Repository  Size
================================================================================
Installing:
 gcc                     x86_64  11.4.1-3.el9                  appstream   32 M
 git-core                x86_64  2.43.0-1.el9                  appstream  4.4 M
 golang-bin              x86_64  1.21.10-1.el9_4               appstream   56 M
 nodejs                  x86_64  1:16.20.2-8.el9_4             appstream  111 k
 npm                     x86_64  1:8.19.4-1.16.20.2.8.el9_4    appstream  1.7 M
 python3-devel           x86_64  3.9.18-3.el9                  appstream  204 k
 python3-pip             noarch  21.2.3-8.el9                  appstream  1.7 M
 python3-setuptools      noarch  53.0.0-12.el9                 baseos     839 k
Upgrading:
 glibc                   x86_64  2.34-100.el9_4.2              baseos     2.0 M
 glibc-common            x86_64  2.34-100.el9_4.2              baseos     301 k
 glibc-minimal-langpack  x86_64  2.34-100.el9_4.2              baseos      27 k
Installing dependencies:
...
```

And this is a snippet of the same build using an older digest (sha256:b2ce50f292ea21d095dce94668a3ca403b8b589923d83fc78def0241a89a2c86):
```
================================================================================
 Package                 Arch    Version                       Repository  Size
================================================================================
Installing:
 gcc                     x86_64  11.4.1-3.el9                  appstream   32 M
 git-core                x86_64  2.43.0-1.el9                  appstream  4.4 M
 golang-bin              x86_64  1.21.10-1.el9_4               appstream   56 M
 nodejs                  x86_64  1:16.20.2-8.el9_4             appstream  111 k
 npm                     x86_64  1:8.19.4-1.16.20.2.8.el9_4    appstream  1.7 M
 python3-devel           x86_64  3.9.18-3.el9                  appstream  204 k
 python3-pip             noarch  21.2.3-8.el9                  appstream  1.7 M
 python3-setuptools      noarch  53.0.0-12.el9                 baseos     839 k
Upgrading:
 glibc                   x86_64  2.34-100.el9_4.2              baseos     2.0 M
 glibc-common            x86_64  2.34-100.el9_4.2              baseos     301 k
 glibc-minimal-langpack  x86_64  2.34-100.el9_4.2              baseos      27 k
 libgcc                  x86_64  11.4.1-3.el9                  baseos      91 k
 libgomp                 x86_64  11.4.1-3.el9                  baseos     267 k
 openssl                 x86_64  1:3.0.7-27.el9                baseos     1.2 M
 openssl-libs            x86_64  1:3.0.7-27.el9                baseos     2.2 M
 python3                 x86_64  3.9.18-3.el9                  baseos      25 k
 python3-libs            x86_64  3.9.18-3.el9                  baseos     7.3 M
Installing dependencies:
...
```

Notice how many more packages are upgraded during the process even though we didn't ask for it in the Containerfile. That is because the metadata database is refreshed every time to the latest one anyway and it's the direct result of package dependencies which we don't have any control over and pinning to a digest clearly isn't going to solve. To top it all off, the ^above is just for our project which installs just a few packages, for bigger projects the difference is going to be much more noticable given that the base set of packages is somewhere around 100+.

Inter-package dependencies/requirements are just one piece of the puzzle. The other is reporting the actual changes inside the container images. With individual packages, you automatically get a changelog that dependabot pulls from the given project which is nice, unless your project depends on some random garbage on github. However, that's not the case with docker images, unless the image vendor doesn't set `org.opencontainers.image.source` label on the image as the [dependabot docs states](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#docker). Now, unlike with individual projects/packages, it is apparently not that common with big reputable vendors out there and platform builder images to set that:
```
$ podman inspect registry.access.redhat.com/ubi9 | grep org.opencontainers.image.source
$ podman inspect docker.io/rockylinux/rockylinux:9-ubi | grep org.opencontainers.image.source
$ podman inspect fedora:39 | grep org.opencontainers.image.source
$ podman inspect docker.io/almalinux/9-base | grep org.opencontainers.image.source
```

And so when you look at the PR dependabot creates for such an image, you're going in blind completely as the only thing you see is a change in an opaque string, but you have literally no idea what has changed - how is that safe (or safer) when the only information you have is whether an image (as the subject of dependabot's PR) passed CI or not? What guarantees do we have about other potential breakages the CI hasn't uncovered yet?

Lastly, @mbasti you mentioned breakages in images and introducing dev versions of packages. I'd say "quite uncommon" and unlikely with platforms like RHEL where Red Hat takes huge pride in specifically not doing that and not introducing any incompatible breakages for the lifetime of the distro and instead keeping it stable for a very long time which is also the reason I even opened this PR and proposed ditching Fedora in the first place, because with Fedora you never know and I've seen broken repo metadata in CentOS Stream preventing package installs as well package upgrades multiple times (which is also something digest pinning will not shield you from at all).

Having said all of ^this and clearly not being a fan of digest pinning with container images despite that being praised as a good practice all over the internet (admittedly though in some cases, e.g. the language toolkit images it does make sense I agree) and I'm not convinced it's really solving anything in our particular case strictly in context of this project as I explained above. However, if you feel so strong about pinning the image for the "perceived" safety and it should be the only blocker for this PR to be accepted, then so be it, but I wanted to express my feelings about it backed by some real life experiment results. Let me know your thoughts so that I can start working on the necessary adjustments.

Reply:
`dnf update` is orthogonal to `:latest`.

The issue here is the (absolutely unfounded) assumption that `:latest` literally means "latest", i.e. the latest *content* (this is perhaps Docker's biggest design/architecture cock-up of *many*).

The *only* meaning of the auto-tag `"latest` is literally "last pushed". You can rely on the HITL to make sure that they're not mistakenly pushing e.g. an F39 image to `F40:latest`, but then you always have to worry about human error. Automation helps, but at that point, why not use a hash, or even better some non-default tag like `:latest-content`?

However: None of that addresses the uncertainty around changes made by `dnf`, but as I said, that's orthogonal, and not isolated to containers - the issue exists with *any* instantiation - VM, bare-metal, container...

Reply:
> `dnf update` is orthogonal to `:latest`.
> 
> The issue here is the (absolutely unfounded) assumption that `:latest` literally means "latest", i.e. the latest _content_ (this is perhaps Docker's biggest design/architecture cock-up of _many_).
> 
> The _only_ meaning of the auto-tag `"latest` is literally "last pushed". You can rely on the HITL to make sure that they're not mistakenly pushing e.g. an F39 image to `F40:latest`, but then you always have to worry about human error. Automation helps, but at that point, why not use a hash, or even better some non-default tag like `:latest-content`?
> 
> However: None of that addresses the uncertainty around changes made by `dnf`, but as I said, that's orthogonal, and not isolated to containers - the issue exists with _any_ instantiation - VM, bare-metal, container...

While I can accept the argument about the "latest push", that would still only apply IMO in random images which one picks as a dependency (just like any other package without a reputable history behind it, typically a package with a single maintainer and contributor). However, we're talking about the official maintained image here, which, I'd like to remind, is just some rebuild of RHEL, so if we don't have trust in maintainers of the image then what can we trust? In that case we shouldn't trust that maintainers don't push errors to RPM repo metadatada which DNF pulls, leading to broken package dependencies. Does it happen? On some distros it sometimes does. Same thing with containers, IMO the important thing here is to have some trust in the well-established community to provide adequate support in case of a problem occurs, like in this case, rebuild the image in a timely manner so that we might not even notice. Again, when we talk about our direct consumables with a massive community and ecosystem rather than "hobby abandonware" we depend on (e.g. reflink which sadly should not have happened) I'm still not seeing how hash pinning is doing us any good.
Since people seem to be in a general consensus of the value hash pinning of docker images brings in our particular case (not in general) despite using reputable sources I guess I can at least sell the idea to myself as a move towards consistency, given that we let dependabot handle the rest of our dependencies.

Reply:
and we shouldn't use latest if possible but stick to minor release version floating tag

Reply:
> [@eskultety](https://github.com/eskultety) I don't get your point
> Why having dependabot gating updates in parent images is worse that blindly consuming anything from the latest? Still limits potential for unexpected breaking change.


Eskultet, I don’t get it either.

I agree with Martin and Ben that we should avoid using :latest at all costs. Dependabot will add an extra CI check for unknown changes in the next UBI version. I am not saying it will prevent us from hitting any issues after moving to the next version but it might catch an issue and IMO it does not make sense not to use it.

Plus removing the linter check is generally a bad practice. 

Maybe I just don’t get the benefit of using the latest here.

Reply:
> CI to be green to merge
this is the reason, CI may catch some issues

:latest tag may cause to bump to next major version and that's usually causes breakages.

Why should I have broken repo when this happens and the rush reactivelly address the issue when I can just prevent it. Saving one cycle of CI per week is nothing.